### ISSUE TITLE
Issue with stateful metrics initialization in ProgbarLogger inside the model.evaluate() function

### ISSUE THREAD

**Issue by evgeniya-egupova** [tensorflow/tensorflow#46713]
**System information**
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from PyPI
- TensorFlow version: 2.4.0
- Python version: 3.6

**Expected behavior**

When using `model.evaluate(validation_dataset)` function, the `ProgbarLogger` callback, that is initialized inside it, should log all the Model's metrics as-is (register them as stateful metrics). 

From `tf.keras.callbacks.ProgbarLogger` docs:
> 
>stateful_metrics | Iterable of string names of metrics that should not be averaged over an epoch. Metrics in this list will be logged as-is. All others will be averaged over time (e.g. loss, etc). If not provided, defaults to the Model's metrics. 
>-- | --

**Behavior in TF 2.3.1**

The `_maybe_init_progbar()` function, which initializes the stateful metrics list, is called inside the `on_test_batch_end()` method. This means that the Model has been evaluated at least once, and therefore the Model's `metrics` attribute is available for the `ProgbarLogger` to initialize the list of stateful metrics.

From `tf.keras.Model` docs:

> Note: metrics_names are available only after a keras.Model has been trained/evaluated on actual data.

**Current behavior in TF 2.4.0**

After b31f069d3abc8446a5d370c6fa65f9df0a7720b6 commit the `_maybe_init_progbar()` function is called inside `on_epoch_begin()`, `on_test_begin()` and `on_predict_begin()`, which means that the Model does not yet contain the list of compiled metrics and the `stateful_metrics` list is not initialized properly. This causes simple metrics such as `tf.keras.metrics.Accuracy` to be averaged over time and then logged to the console.

**Comment by ravikyram** [comment]
@evgeniya-egupova 

Can you please share colab link or simple standalone code to reproduce the issue in our environment.It helps us in localizing the issue faster. Thanks!

**Comment by axeldavy** [comment]
I have an issue that looks related:

Since using 2.4.1 (I used 2.3.0 before), during training the values reported for my metrics seem to be incorrect. For example the "loss" reported is different than the one passed to 'on_epoch_end'. Maybe there is double averaging going on.

**Comment by evgeniya-egupova** [comment]
@ravikyram 

Here is [the gist showing the tf.keras.applications.MobileNetV2 evaluation](https://gist.github.com/evgeniya-egupova/f50bb4e5ec37e152c1992db965bc04ab). The accuracy on imagenet is 71.85%, however the logged accuracy is 72.03%.

Note that there are some evaluation issues with batch size > 1, so I suggest not to change it.

**Comment by ssomers** [comment]
Allow me to summarize the bug a bit more accurately: tensorflow writes bogus values for training metrics on the console if verbose=1 [as is the default](https://www.tensorflow.org/api_docs/python/tf/keras/Model#fit)). The console output is fine with verbose=2. Regardless of that parameter, the correct values are logged in history (and to tensorboard, when requested). Validation metrics are unaffected.

So a workaround is to set verbose=2, or simply to continue enjoying the dancing numbers but ignoring them when the dance is over.

[Here is a much simpler demonstration](https://gist.github.com/ssomers/59c16f042c8c356a442636492f258cc2).

**Comment by ricvo** [comment]
@ssomers it's true the problem seems to be in the Progbar

Related issue https://github.com/tensorflow/tensorflow/issues/47099 has been closed, could you check if the present issue is solved too on the new commit 979c5093da6238eae2f137e2f5538cf6373c4c02?

**Comment by ssomers** [comment]
The fresh [tf-nightly 2.6.0.dev20210421](https://pypi.org/project/tf-nightly/) doesn't help for this issue, but I can't find which commit this is built on (and meanwhile it seems the Windows wheels have disappeared). Going over to trying to build tf myself.

**Comment by ssomers** [comment]
I'm pretty sure I tested the wrong thing yesterday. Today, a simple `pip install tf-nightly` in a fresh python 3.9 virtual environment ends in a flawless display of loss and custom metrics.

I'm not so sure that that means this issue is entirely fixed and is the same as all the linked issues.

**Comment by ymodak** [comment]
This is fixed with latest tf-nightly as mentioned by ssomers.
See [gist](https://colab.research.google.com/gist/jvishnuvardhan/ab673604087e19dc352b131c550992e8/untitled101.ipynb) for your reference. Thanks!

**Comment by alexander-soare** [comment]
@ymodak I am on tf-nightly-2.7.0.dev20210630 and am still observing that stateful metrics are not as described (ie, they continue to be averaged over the running epoch)

**Comment by chunduriv** [comment]
@alexander-soare, 

Is it still an issue for you? Please feel free to open a new issue. Thank you.
